src/librustc/mir/mod.rs

```diff
@@ -2417,6 +2425,10 @@ impl<'tcx> TypeFoldable<'tcx> for Rvalue<'tcx> {
             }
         }
     }
+
+    fn super_hash_with<H: TypeHasher<'tcx>>(&self, _hasher: &mut H) -> u64 {
+        unimplemented!()
+    }
 }
 
 impl<'tcx> TypeFoldable<'tcx> for Operand<'tcx> {
```

Comment:
I wonder if these impls can be moved to the macros -- I moved a bunch of stuff in preparation for this PR, but not everything. Trying to move as many things we can to use the macro -- before we make any changes to it! -- might indeed be a good first step.

Reply:
Ok, I've made a start on this - I've picked out a couple that seem to be trivially applicable to the macros. For a couple of the types, it seems that only some of the fields (in the case of structs) and components of variants (in the case of enums) are `super_fold`ed or `super_visit`ed. Would an approach be to somehow 'tag' these in the macro and extend the macro to understand the difference?

I'm imagining for example the impl for `RValue`, which does something like:
```rust
        match *self {
            Use(ref op) => Use(op.fold_with(folder)),
            Repeat(ref op, len) => Repeat(op.fold_with(folder), len),
            Ref(region, bk, ref place) =>
                Ref(region.fold_with(folder), bk, place.fold_with(folder)),
            Len(ref place) => Len(place.fold_with(folder)),
*snip*
       match *self {
            Use(ref op) => op.visit_with(visitor),
            Repeat(ref op, _) => op.visit_with(visitor),
            Ref(region, _, ref place) => region.visit_with(visitor) || place.visit_with(visitor),
            Len(ref place) => place.visit_with(visitor),
```

and the macro could be something like:

```
EnumTypeFoldableImpl! {
    impl<'tcx, T> TypeFoldable<'tcx> for Rvalue<'tcx> {
        (Rvalue::Use)(op),
        (Rvalue::Repeat)(op, !clone len),
        (Rvalue::Ref)(region, !clone bk, place),
        (Rvalue::Len)(place),
        ...
    }
}
````